x-pack/elastic-agent/docs/elastic-agent-dynamic-inputs.asciidoc

```diff
@@ -0,0 +1,837 @@
+would use `{{foo.key1}}` and `{{foo.key2}}`.
+
+=== Provider configuration
+
+The provider configuration is specified under the top-level `providers`
+key in the `elastic-agent.yml` configuration. By default, all registered
+providers are enabled. If a provider cannot connect, it produces no mappings.
+
+The following examples shows two providers, `local` and `local_dynamic`, that
+supply custom keys:
+
+[source,yaml]
+----
+providers:
+  local:
+    vars:
+      foo: bar
+  local_dynamic:
+    vars:
+      - item: key1
+      - item: key2
+----
+
+To explicitly disable a provider, set `enabled: false`. Because all providers
+are prefixed and have no name collisions, the name of the provider is the key in
+the configuration.
+
+[source,yaml]
+----
+providers:
+  docker:
+    enabled: false
+----
+
+{agent} supports two broad types of providers: <<context-providers,context>> and
+<<dynamic-providers,dynamic>>.
+
+[[context-providers]]
+=== Context providers
+
+Context providers provide the current context of the running {agent}, for
+example, agent information (id, version), host information (hostname, IP
+addresses), and environment information (environment variables).
+
+They can only provide a single key/value mapping. Think of them as singletons;
+an update of a key/value mapping will result in a re-evaluation of the entire
+configuration. These providers are normally very static, but that's not
+required. It is possible for a value to change resulting in re-evaluation.
+
+Context providers use ECS naming when possible to ensure that documentation and
+understanding across projects is the same.
+
+{agent} supports the following context providers:
+
+[[local-provider]]
+==== Local
+
+Provides custom keys to use as variables. For example:
+
+[source,yaml]
+----
+providers:
+  local:
+    vars:
+      foo: bar
+----
+
+[[agent-provider]]
+==== Agent provider
+
+Provides information about the {agent}. The available keys are:
+
+//PH: I think it might be easier for users if we show the keys they can use
+//in their config. Is this correct? 
+
+`agent.id`:: Current agent ID
+`agent.version`:: Current agent version information object
+`agent.version.version`:: Current agent version
+`agent.version.commit`:: Version commit
+`agent.version.build_time`:: Version build time
+`agent.version.snapshot`:: Version is snapshot build
+
+//PH: Is snapshot a boolean? Should we document the types here? 
+
+[[host-provider]]
+==== Host provider
+
+Provides information about the current host. The available keys are:
+
+`host.name`:: Host name
+`host.platform`:: Host platform
+`host.architecture`:: Host architecture
+`host.ip[]`:: Host IP addresses
+`host.mac[]`:: Host MAC addresses
+
+[[env-provider]]
+==== Env Provider
+
+Provides access to the environment variables as key/values.
+
+//PH: Maybe we should provide some concrete examples here? To access env var
+//would you need to use something like `${env.SHELL}`??
```

Comment:
The problem is a bit complicated because you actually need to edit the starting scripts.

```
foo=bar elastic-agent run 
````
And be used with 
${env.foo}